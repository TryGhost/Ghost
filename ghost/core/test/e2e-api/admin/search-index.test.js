const {agentProvider, fixtureManager, matchers} = require('../../utils/e2e-framework');
const {anyContentVersion, anyEtag, anyISODateTime, anyString} = matchers;
const assert = require('node:assert');

describe('Search Index API', function () {
    let agent;

    before(async function () {
        agent = await agentProvider.getAdminAPIAgent();
        await fixtureManager.init('posts');
        await agent.loginAsOwner();
    });

    describe('fetchPosts', function () {
        const searchIndexPostMatcher = {
            id: anyString,
            title: anyString,
            url: anyString,
            status: anyString,
            published_at: anyISODateTime,
            visibility: anyString
        };

        it('should return a list of posts', async function () {
            const response = await agent.get('/search-index/posts')
                .expectStatus(200)
                .matchHeaderSnapshot({
                    'content-version': anyContentVersion,
                    etag: anyEtag
                })
                .matchBodySnapshot({
                    posts: new Array(11).fill(searchIndexPostMatcher)
                });

            // Explicitly double-check that expensive fields are not included
            const post = response.body.posts[0];
            assert.equal(post.excerpt, undefined);
            assert.equal(post.html, undefined);
            assert.equal(post.mobiledoc, undefined);
            assert.equal(post.lexical, undefined);
            assert.equal(post.plaintext, undefined);
        });
    });

    describe('fetchPages', function () {
        const searchIndexPageMatcher = {
            id: anyString,
            title: anyString,
            url: anyString,
            status: anyString,
            published_at: anyISODateTime,
            visibility: anyString
        };

        it('should return a list of pages', async function () {
            const response = await agent.get('/search-index/pages')
                .expectStatus(200)
                .matchHeaderSnapshot({
                    'content-version': anyContentVersion,
                    etag: anyEtag
                })
                .matchBodySnapshot({
                    pages: new Array(5).fill(searchIndexPageMatcher)
                });

            // Explicitly double-check that expensive fields are not included
            const page = response.body.pages[0];
            assert.equal(page.excerpt, undefined);
            assert.equal(page.html, undefined);
            assert.equal(page.mobiledoc, undefined);
            assert.equal(page.lexical, undefined);
            assert.equal(page.plaintext, undefined);
        });
    });

    describe('fetchTags', function () {
        const searchIndexTagMatcher = {
            id: anyString,
            slug: anyString,
            name: anyString,
            url: anyString
        };

        it('should return a list of tags', async function () {
            await agent.get('/search-index/tags')
                .expectStatus(200)
                .matchHeaderSnapshot({
                    'content-version': anyContentVersion,
                    etag: anyEtag
                })
                .matchBodySnapshot({
                    tags: new Array(6).fill(searchIndexTagMatcher)
                });
        });
    });

    describe('fetchUsers', function () {
        const searchIndexUserMatcher = {
            id: anyString,
            slug: anyString,
            name: anyString,
            url: anyString,
            profile_image: anyString
        };

        it('should return a list of users', async function () {
            await agent.get('/search-index/users')
                .expectStatus(200)
                .matchHeaderSnapshot({
                    'content-version': anyContentVersion,
                    etag: anyEtag
                })
                .matchBodySnapshot({
                    users: new Array(2).fill(searchIndexUserMatcher)
                });
        });
    });
});
