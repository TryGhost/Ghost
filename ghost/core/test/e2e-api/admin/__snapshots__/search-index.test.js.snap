// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Search Index API fetchPages should return a list of pages 1: [body] 1`] = `
Object {
  "pages": Array [
    Object {
      "id": Any<String>,
      "published_at": StringMatching /\\\\d\\{4\\}-\\\\d\\{2\\}-\\\\d\\{2\\}T\\\\d\\{2\\}:\\\\d\\{2\\}:\\\\d\\{2\\}\\\\\\.000Z/,
      "status": Any<String>,
      "title": Any<String>,
      "url": Any<String>,
      "visibility": Any<String>,
    },
    Object {
      "id": Any<String>,
      "published_at": StringMatching /\\\\d\\{4\\}-\\\\d\\{2\\}-\\\\d\\{2\\}T\\\\d\\{2\\}:\\\\d\\{2\\}:\\\\d\\{2\\}\\\\\\.000Z/,
      "status": Any<String>,
      "title": Any<String>,
      "url": Any<String>,
      "visibility": Any<String>,
    },
    Object {
      "id": Any<String>,
      "published_at": StringMatching /\\\\d\\{4\\}-\\\\d\\{2\\}-\\\\d\\{2\\}T\\\\d\\{2\\}:\\\\d\\{2\\}:\\\\d\\{2\\}\\\\\\.000Z/,
      "status": Any<String>,
      "title": Any<String>,
      "url": Any<String>,
      "visibility": Any<String>,
    },
    Object {
      "id": Any<String>,
      "published_at": StringMatching /\\\\d\\{4\\}-\\\\d\\{2\\}-\\\\d\\{2\\}T\\\\d\\{2\\}:\\\\d\\{2\\}:\\\\d\\{2\\}\\\\\\.000Z/,
      "status": Any<String>,
      "title": Any<String>,
      "url": Any<String>,
      "visibility": Any<String>,
    },
    Object {
      "id": Any<String>,
      "published_at": StringMatching /\\\\d\\{4\\}-\\\\d\\{2\\}-\\\\d\\{2\\}T\\\\d\\{2\\}:\\\\d\\{2\\}:\\\\d\\{2\\}\\\\\\.000Z/,
      "status": Any<String>,
      "title": Any<String>,
      "url": Any<String>,
      "visibility": Any<String>,
    },
  ],
}
`;

exports[`Search Index API fetchPages should return a list of pages 2: [headers] 1`] = `
Object {
  "access-control-allow-origin": "http://127.0.0.1:2369",
  "cache-control": "no-cache, private, no-store, must-revalidate, max-stale=0, post-check=0, pre-check=0",
  "content-length": "926",
  "content-type": "application/json; charset=utf-8",
  "content-version": StringMatching /v\\\\d\\+\\\\\\.\\\\d\\+/,
  "etag": StringMatching /\\(\\?:W\\\\/\\)\\?"\\(\\?:\\[ !#-\\\\x7E\\\\x80-\\\\xFF\\]\\*\\|\\\\r\\\\n\\[\\\\t \\]\\|\\\\\\\\\\.\\)\\*"/,
  "vary": "Accept-Version, Origin, Accept-Encoding",
  "x-powered-by": "Express",
}
`;

exports[`Search Index API fetchPosts should return a list of posts 1: [body] 1`] = `
Object {
  "posts": Array [
    Object {
      "id": Any<String>,
      "published_at": StringMatching /\\\\d\\{4\\}-\\\\d\\{2\\}-\\\\d\\{2\\}T\\\\d\\{2\\}:\\\\d\\{2\\}:\\\\d\\{2\\}\\\\\\.000Z/,
      "status": Any<String>,
      "title": Any<String>,
      "url": Any<String>,
      "visibility": Any<String>,
    },
    Object {
      "id": Any<String>,
      "published_at": StringMatching /\\\\d\\{4\\}-\\\\d\\{2\\}-\\\\d\\{2\\}T\\\\d\\{2\\}:\\\\d\\{2\\}:\\\\d\\{2\\}\\\\\\.000Z/,
      "status": Any<String>,
      "title": Any<String>,
      "url": Any<String>,
      "visibility": Any<String>,
    },
    Object {
      "id": Any<String>,
      "published_at": StringMatching /\\\\d\\{4\\}-\\\\d\\{2\\}-\\\\d\\{2\\}T\\\\d\\{2\\}:\\\\d\\{2\\}:\\\\d\\{2\\}\\\\\\.000Z/,
      "status": Any<String>,
      "title": Any<String>,
      "url": Any<String>,
      "visibility": Any<String>,
    },
    Object {
      "id": Any<String>,
      "published_at": StringMatching /\\\\d\\{4\\}-\\\\d\\{2\\}-\\\\d\\{2\\}T\\\\d\\{2\\}:\\\\d\\{2\\}:\\\\d\\{2\\}\\\\\\.000Z/,
      "status": Any<String>,
      "title": Any<String>,
      "url": Any<String>,
      "visibility": Any<String>,
    },
    Object {
      "id": Any<String>,
      "published_at": StringMatching /\\\\d\\{4\\}-\\\\d\\{2\\}-\\\\d\\{2\\}T\\\\d\\{2\\}:\\\\d\\{2\\}:\\\\d\\{2\\}\\\\\\.000Z/,
      "status": Any<String>,
      "title": Any<String>,
      "url": Any<String>,
      "visibility": Any<String>,
    },
    Object {
      "id": Any<String>,
      "published_at": StringMatching /\\\\d\\{4\\}-\\\\d\\{2\\}-\\\\d\\{2\\}T\\\\d\\{2\\}:\\\\d\\{2\\}:\\\\d\\{2\\}\\\\\\.000Z/,
      "status": Any<String>,
      "title": Any<String>,
      "url": Any<String>,
      "visibility": Any<String>,
    },
    Object {
      "id": Any<String>,
      "published_at": StringMatching /\\\\d\\{4\\}-\\\\d\\{2\\}-\\\\d\\{2\\}T\\\\d\\{2\\}:\\\\d\\{2\\}:\\\\d\\{2\\}\\\\\\.000Z/,
      "status": Any<String>,
      "title": Any<String>,
      "url": Any<String>,
      "visibility": Any<String>,
    },
    Object {
      "id": Any<String>,
      "published_at": StringMatching /\\\\d\\{4\\}-\\\\d\\{2\\}-\\\\d\\{2\\}T\\\\d\\{2\\}:\\\\d\\{2\\}:\\\\d\\{2\\}\\\\\\.000Z/,
      "status": Any<String>,
      "title": Any<String>,
      "url": Any<String>,
      "visibility": Any<String>,
    },
    Object {
      "id": Any<String>,
      "published_at": StringMatching /\\\\d\\{4\\}-\\\\d\\{2\\}-\\\\d\\{2\\}T\\\\d\\{2\\}:\\\\d\\{2\\}:\\\\d\\{2\\}\\\\\\.000Z/,
      "status": Any<String>,
      "title": Any<String>,
      "url": Any<String>,
      "visibility": Any<String>,
    },
    Object {
      "id": Any<String>,
      "published_at": StringMatching /\\\\d\\{4\\}-\\\\d\\{2\\}-\\\\d\\{2\\}T\\\\d\\{2\\}:\\\\d\\{2\\}:\\\\d\\{2\\}\\\\\\.000Z/,
      "status": Any<String>,
      "title": Any<String>,
      "url": Any<String>,
      "visibility": Any<String>,
    },
    Object {
      "id": Any<String>,
      "published_at": StringMatching /\\\\d\\{4\\}-\\\\d\\{2\\}-\\\\d\\{2\\}T\\\\d\\{2\\}:\\\\d\\{2\\}:\\\\d\\{2\\}\\\\\\.000Z/,
      "status": Any<String>,
      "title": Any<String>,
      "url": Any<String>,
      "visibility": Any<String>,
    },
  ],
}
`;

exports[`Search Index API fetchPosts should return a list of posts 1: [headers] 1`] = `
Object {
  "access-control-allow-origin": "http://127.0.0.1:2369",
  "cache-control": "no-cache, private, no-store, must-revalidate, max-stale=0, post-check=0, pre-check=0",
  "content-length": "106403",
  "content-type": "application/json; charset=utf-8",
  "content-version": StringMatching /v\\\\d\\+\\\\\\.\\\\d\\+/,
  "etag": StringMatching /\\(\\?:W\\\\/\\)\\?"\\(\\?:\\[ !#-\\\\x7E\\\\x80-\\\\xFF\\]\\*\\|\\\\r\\\\n\\[\\\\t \\]\\|\\\\\\\\\\.\\)\\*"/,
  "vary": "Accept-Version, Origin, Accept-Encoding",
  "x-powered-by": "Express",
}
`;

exports[`Search Index API fetchPosts should return a list of posts 2: [headers] 1`] = `
Object {
  "access-control-allow-origin": "http://127.0.0.1:2369",
  "cache-control": "no-cache, private, no-store, must-revalidate, max-stale=0, post-check=0, pre-check=0",
  "content-length": "2313",
  "content-type": "application/json; charset=utf-8",
  "content-version": StringMatching /v\\\\d\\+\\\\\\.\\\\d\\+/,
  "etag": StringMatching /\\(\\?:W\\\\/\\)\\?"\\(\\?:\\[ !#-\\\\x7E\\\\x80-\\\\xFF\\]\\*\\|\\\\r\\\\n\\[\\\\t \\]\\|\\\\\\\\\\.\\)\\*"/,
  "vary": "Accept-Version, Origin, Accept-Encoding",
  "x-powered-by": "Express",
}
`;

exports[`Search Index API fetchTags should return a list of tags 1: [body] 1`] = `
Object {
  "tags": Array [
    Object {
      "id": Any<String>,
      "name": Any<String>,
      "slug": Any<String>,
      "url": Any<String>,
    },
    Object {
      "id": Any<String>,
      "name": Any<String>,
      "slug": Any<String>,
      "url": Any<String>,
    },
    Object {
      "id": Any<String>,
      "name": Any<String>,
      "slug": Any<String>,
      "url": Any<String>,
    },
    Object {
      "id": Any<String>,
      "name": Any<String>,
      "slug": Any<String>,
      "url": Any<String>,
    },
    Object {
      "id": Any<String>,
      "name": Any<String>,
      "slug": Any<String>,
      "url": Any<String>,
    },
    Object {
      "id": Any<String>,
      "name": Any<String>,
      "slug": Any<String>,
      "url": Any<String>,
    },
  ],
}
`;

exports[`Search Index API fetchTags should return a list of tags 1: [headers] 1`] = `
Object {
  "access-control-allow-origin": "http://127.0.0.1:2369",
  "cache-control": "no-cache, private, no-store, must-revalidate, max-stale=0, post-check=0, pre-check=0",
  "content-length": "693",
  "content-type": "application/json; charset=utf-8",
  "content-version": StringMatching /v\\\\d\\+\\\\\\.\\\\d\\+/,
  "etag": StringMatching /\\(\\?:W\\\\/\\)\\?"\\(\\?:\\[ !#-\\\\x7E\\\\x80-\\\\xFF\\]\\*\\|\\\\r\\\\n\\[\\\\t \\]\\|\\\\\\\\\\.\\)\\*"/,
  "vary": "Accept-Version, Origin, Accept-Encoding",
  "x-powered-by": "Express",
}
`;

exports[`Search Index API fetchTags should return a list of tags 2: [headers] 1`] = `
Object {
  "access-control-allow-origin": "http://127.0.0.1:2369",
  "cache-control": "no-cache, private, no-store, must-revalidate, max-stale=0, post-check=0, pre-check=0",
  "content-length": "693",
  "content-type": "application/json; charset=utf-8",
  "content-version": StringMatching /v\\\\d\\+\\\\\\.\\\\d\\+/,
  "etag": StringMatching /\\(\\?:W\\\\/\\)\\?"\\(\\?:\\[ !#-\\\\x7E\\\\x80-\\\\xFF\\]\\*\\|\\\\r\\\\n\\[\\\\t \\]\\|\\\\\\\\\\.\\)\\*"/,
  "vary": "Accept-Version, Origin, Accept-Encoding",
  "x-powered-by": "Express",
}
`;

exports[`Search Index API fetchUsers should return a list of users 1: [body] 1`] = `
Object {
  "users": Array [
    Object {
      "id": Any<String>,
      "name": Any<String>,
      "profile_image": Any<String>,
      "slug": Any<String>,
      "url": Any<String>,
    },
    Object {
      "id": Any<String>,
      "name": Any<String>,
      "profile_image": Any<String>,
      "slug": Any<String>,
      "url": Any<String>,
    },
  ],
}
`;

exports[`Search Index API fetchUsers should return a list of users 2: [headers] 1`] = `
Object {
  "access-control-allow-origin": "http://127.0.0.1:2369",
  "cache-control": "no-cache, private, no-store, must-revalidate, max-stale=0, post-check=0, pre-check=0",
  "content-length": "368",
  "content-type": "application/json; charset=utf-8",
  "content-version": StringMatching /v\\\\d\\+\\\\\\.\\\\d\\+/,
  "etag": StringMatching /\\(\\?:W\\\\/\\)\\?"\\(\\?:\\[ !#-\\\\x7E\\\\x80-\\\\xFF\\]\\*\\|\\\\r\\\\n\\[\\\\t \\]\\|\\\\\\\\\\.\\)\\*"/,
  "vary": "Accept-Version, Origin, Accept-Encoding",
  "x-powered-by": "Express",
}
`;
